File: R-CNN/Bilinear_Interpolation/BI.py
import matplotlib.pyplot as plt
from matplotlib.image import imread
import numpy as np
import math

# 使用，双线性插值，来，放大或缩小图像
def BiLinear_interpolation(img, dstH, dstW):
    scrH, scrW, _ = img.shape
    print(img.shape)
    img = np.pad(img, ((0, 1), (0, 1), (0, 0)), 'constant')
    print(img.shape)
    retimg = np.zeros((dstH, dstW, 3), dtype=np.uint8)
    for i in range(dstH):
        for j in range(dstW):
            scrx = (i+0.5)*(scrH/dstH)-0.5
            scry = (j+0.5)*(scrW/dstW)-0.5
            x = math.floor(scrx)
            y = math.floor(scry)
            u = scrx-x
            v = scry-y
            
            # retimg[i,j]=(1-u)*(1-v)*img[x,y]+u*(1-v)*img[x+1,y]+(1-u)*v*img[x,y+1]+u*v*img[x+1,y+1]
            
            v1 = (1-u)*(1-v)
            v2 = u*(1-v)
            v3 = (1-u)*v
            v4 = u*v

            item1 = v1*img[x, y]    
            item2 = v2*img[x+1, y]
            item3 = v3*img[x, y+1]
            item4 = v4*img[x+1, y+1]

            retimg[i, j] = item1+item2+item3+item4
    return retimg


image = imread('test.jpg')

# image = image*255 //.png需要乘255

# image2 = BiLinear_interpolation(image, image.shape[0]*10, image.shape[1]*10)  # 放大

# image2 = BiLinear_interpolation(image, math.floor(image.shape[0]*0.5), math.floor(image.shape[1]*0.5))   # 缩小

image2 = BiLinear_interpolation(image, math.floor(image.shape[0]*0.5), math.floor(image.shape[1]*1))

plt.imshow(image2)
plt.show()


# https://blog.csdn.net/qq_41076797/article/details/114546796

File: R-CNN/Bilinear_Interpolation/BI.js
const np = require('../../Deep-learning/jsStudyNote1/common/numpy')
const { imread, imshow } = require('../../Deep-learning/jsStudyNote1/common/matplotlib')

function BiLinear_interpolation(img, dstH, dstW) {
    const [scrH, scrW, _] = np.shape(img)
    img = np.pad(img, [[0, 1], [0, 1], [0, 0]], 'constant')
    let retimg = np.zeros([dstH, dstW, 3])
    for (let i = 0; i < dstH; i++) {
        for (let j = 0; j < dstW; j++) {
            let scrx = (i + 0.5) * (scrH / dstH) - 0.5
            let scry = (j + 0.5) * (scrW / dstW) - 0.5
            let x = Math.floor(scrx)
            let y = Math.floor(scry)
            let u = scrx - x
            let v = scry - y

            //快
            let v1 = (1 - u) * (1 - v)
            let v2 = u * (1 - v)
            let v3 = (1 - u) * v
            let v4 = u * v
            let item1 = np.expr(v1, '*', img[x][y])
            let item2 = np.expr(v2, '*', img[x + 1][y])
            let item3 = np.expr(v3, '*', img[x][y + 1])
            let item4 = np.expr(v4, '*', img[x + 1][y + 1])
            let value = np.expr(item1, '+', np.expr(item2, '+', np.expr(item3, '+', item4)))
            retimg[i][j] = value;

            // //慢
            // let item1 = np.expr((1 - u) * (1 - v), '*', np.indexing(img, [x, y]))
            // let item2 = np.expr(u * (1 - v), '*', np.indexing(img, [x + 1, y]))
            // let item3 = np.expr((1 - u) * v, '*', np.indexing(img, [x, y + 1]))
            // let item4 = np.expr(u * v, '*', np.indexing(img, [x + 1, y + 1]))
            // np.indexing(retimg, [i, j], np.expr(item1, '+', np.expr(item2, '+', np.expr(item3, '+', item4))))

            console.log([i, j])
        }
    }
    return retimg
}

imread('./test.jpg', (imageData) => {
    console.log(imageData)

    const { width, height, data } = imageData

    image = np.reshape(data, [height, width, 4])

    image = np.indexing(image, [np.slice(np.None), np.slice(np.None), np.slice(0, 3)])

    // image2 = BiLinear_interpolation(image, Math.floor(height * 0.5), Math.floor(width * 1))

    // image2 = BiLinear_interpolation(image, Math.floor(height * 0.5), Math.floor(width * 0.5))

    imshow(image2)
})